// atool-build/lib/build.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (args, callback) {
  args.cwd = args.cwd || process.cwd();

  // Get config.
  var webpackConfig = getWebpackConfig(args);

  // Clean output dir first.
  _rimraf2.default.sync(webpackConfig.output.path);

  function doneHandler(err, stats) {
    var _stats$toJson = stats.toJson();

    var errors = _stats$toJson.errors;

    if (errors && errors.length) {
      process.on('exit', function exitHandler() {
        process.exit(1);
      });
    }

    console.log(stats.toString({ colors: true }));

    if (err) {
      process.on('exit', function exitHandler() {
        process.exit(1);
      });
      console.error(err);
    }
    if (callback) {
      callback(err);
    }
  }

  // Run compiler.
  var compiler = (0, _webpack2.default)(webpackConfig);
  if (args.watch) {
    compiler.watch(args.watch || 200, doneHandler);
  } else {
    compiler.run(doneHandler);
  }
};

var _path = require('path');

var _rimraf = require('rimraf');

var _rimraf2 = _interopRequireDefault(_rimraf);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _mergeCustomConfig = require('./mergeCustomConfig');

var _mergeCustomConfig2 = _interopRequireDefault(_mergeCustomConfig);

var _getWebpackCommonConfig = require('./getWebpackCommonConfig');

var _getWebpackCommonConfig2 = _interopRequireDefault(_getWebpackCommonConfig);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function getWebpackConfig(args) {
  var webpackConfig = (0, _getWebpackCommonConfig2.default)(args);

  webpackConfig.plugins = webpackConfig.plugins || [];

  // Config outputPath.
  if (args.outputPath) {
    webpackConfig.output.path = args.outputPath;
  }

  // Config if no --no-compress.
  if (args.compress) {
    webpackConfig.plugins = [].concat(_toConsumableArray(webpackConfig.plugins), [new _webpack2.default.optimize.UglifyJsPlugin({
      output: {
        ascii_only: true
      },
      compress: {
        warnings: false
      }
    })]);
  }

  webpackConfig.plugins = [].concat(_toConsumableArray(webpackConfig.plugins), [new _webpack2.default.DefinePlugin({
    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || 'production')
  })]);

  // Output map.json if hash.
  if (args.hash) {
    var pkg = require((0, _path.join)(args.cwd, 'package.json'));
    webpackConfig.output.filename = webpackConfig.output.chunkFilename = '[name]-[chunkhash].js';
    webpackConfig.plugins = [].concat(_toConsumableArray(webpackConfig.plugins), [require('map-json-webpack-plugin')({
      assetsPath: pkg.name,
      output: (0, _path.join)(webpackConfig.output.path, 'map.json')
    })]);
  }

  webpackConfig = (0, _mergeCustomConfig2.default)(webpackConfig, (0, _path.join)(args.cwd, args.config || 'webpack.config.js'));

  return webpackConfig;
}

module.exports = exports['default'];
