// dora/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createServer;

var _http = require('http');

var _http2 = _interopRequireDefault(_http);

var _koa = require('koa');

var _koa2 = _interopRequireDefault(_koa);

var _plugin = require('./plugin');

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _spmLog = require('spm-log');

var _spmLog2 = _interopRequireDefault(_spmLog);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultCwd = process.cwd();
var defaultArgs = {
  port: '8000',
  cwd: defaultCwd,
  resolveDir: [defaultCwd]
};

function createServer(_args) {
  var args = (0, _objectAssign2.default)({}, defaultArgs, _args);
  _spmLog2.default.config(args);

  var pluginNames = args.plugins;
  var port = args.port;
  var cwd = args.cwd;

  var pluginArgs = {
    port: port,
    cwd: cwd
  };
  var plugins = (0, _plugin.resolvePlugins)(pluginNames, args.resolveDir, args.cwd);
  function _applyPlugins(name, applyArgs) {
    return (0, _plugin.applyPlugins)(plugins, name, pluginArgs, applyArgs);
  }
  pluginArgs.applyPlugins = _applyPlugins;
  _spmLog2.default.debug('dora', '[plugins] ' + JSON.stringify(plugins));
  var app = (0, _koa2.default)();

  _applyPlugins('middleware.before');
  (0, _plugin.applyMiddlewares)(plugins, pluginArgs, app);
  app.use(require('koa-static-with-post')(cwd));
  app.use(require('koa-serve-index')(cwd, {
    hidden: true,
    view: 'details'
  }));
  _applyPlugins('middleware.after');

  _applyPlugins('server.before');
  var server = _http2.default.createServer(app.callback());
  server.listen(port, function () {
    // Fix log, #8
    var stream = process.stderr;
    if (stream.isTTY) {
      stream.cursorTo(0);
      stream.clearLine(1);
    }

    _spmLog2.default.info('dora', 'listened on ' + port);
    _applyPlugins('server.after');
  });
}
module.exports = exports['default'];
