// dora-plugin-proxy/lib/utils.js
'use strict';

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; })();

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isRemote = isRemote;
exports.isMatch = isMatch;
exports.getParams = getParams;
exports.getRes = getRes;

var _url = require('url');

var _pathToRegexp = require('path-to-regexp');

var _pathToRegexp2 = _interopRequireDefault(_pathToRegexp);

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var _isPlainObject = require('is-plain-object');

var _isPlainObject2 = _interopRequireDefault(_isPlainObject);

var _withDb = require('mime-type/with-db');

var _withDb2 = _interopRequireDefault(_withDb);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj; }

function decodeParam(val) {
  if (typeof val !== 'string' || val.length === 0) {
    return val;
  }

  return decodeURIComponent(val);
}

function isRemote(str) {
  return str.indexOf('http://') === 0 || str.indexOf('https://') === 0;
}

function isMatch(req, pattern) {
  var method = req.method;
  var url = req.url;

  var urlObj = (0, _url.parse)(url);
  var expectPattern = undefined;
  var expectMethod = undefined;

  if (pattern.indexOf(' ') > -1) {
    var _pattern$split = pattern.split(/\s+/);

    var _pattern$split2 = _slicedToArray(_pattern$split, 2);

    expectMethod = _pattern$split2[0];
    expectPattern = _pattern$split2[1];
  } else {
    expectPattern = pattern;
  }
  // Match method first.
  if (expectMethod && expectMethod.toUpperCase() !== method.toUpperCase()) {
    return false;
  }

  if (isRemote(expectPattern)) {
    var _parseUrl = (0, _url.parse)(expectPattern);

    var hostname = _parseUrl.hostname;
    var port = _parseUrl.port;
    var path = _parseUrl.path;

    return hostname === urlObj.hostname && (port || '80') === (urlObj.port || '80') && !!urlObj.path.match((0, _pathToRegexp2.default)(path));
  }

  return !!urlObj.pathname.match((0, _pathToRegexp2.default)(expectPattern));
}

function getParams(url, pattern) {
  var keys = [];
  var path = pattern.trim().indexOf(' ') > -1 ? pattern.split(' ')[1] : pattern;
  if (path === '/') {
    return {};
  }

  var regexp = (0, _pathToRegexp2.default)(path, keys);
  var m = regexp.exec(url);
  if (!keys.length || !m) {
    return {};
  }

  var params = {};
  m.forEach(function (ms, index) {
    if (index === 0) return;
    var key = keys[index - 1];
    var prop = key.name;
    var val = decodeParam(ms);
    if (val !== undefined || !Object.prototype.hasOwnProperty.call(params, prop)) {
      params[prop] = val;
    }
  });
  return params;
}

function getRes(req, callback) {
  var _status = 200;
  var headers = {};

  function normalizeData(data) {
    switch (typeof data === 'undefined' ? 'undefined' : _typeof(data)) {
      case 'string':
        return data;
      default:
        return JSON.stringify(data);
    }
  }

  return {
    type: function type(_type) {
      return this.set('Content-Type', _withDb2.default.lookup(_type));
    },
    set: function set(key, val) {
      if ((0, _isPlainObject2.default)(key)) {
        (0, _objectAssign2.default)(headers, key);
      } else {
        headers[key] = val;
      }
      return this;
    },
    status: function status(statusCode) {
      _status = statusCode;
      return this;
    },
    json: function json(data) {
      return this.type('json').end(JSON.stringify(data));
    },
    jsonp: function jsonp(data, callbackName) {
      var fn = req.query[callbackName || 'callback'];
      return this.type('json').end(fn + '(' + JSON.stringify(data) + ')');
    },
    end: function end(data) {
      callback(_status, headers, normalizeData(data));
      return this;
    }
  };
}
